hppc-template-intrinsics/src/main/java/com/carrotsearch/hppc/Intrinsics.java

```diff
@@ -137,4 +137,21 @@ public static <T> T add(T op1, T op2) {
 
     throw new UnsupportedOperationException("Invalid for arbitrary types: " + op1 + " " + op2);
   }
+
+  /**
+   * Returns a numerical value for the argument for primitive template types. This intrinsic is used
+   * to apply arithmetic operations on keys. It is invalid for generic types.
+   */
+  public static <T> int numeric(T e) {
```

Comment:
I add this intrinsic "numeric" to be able to compare numeric values (greater/less than) and also to compute the numeric slopes and intercepts at the core of the index learning.

Reply:
I haven't had the time to read through the paper yet so I may be saying something stupid here but why does it return the int value for everything, including long/ floating point values? Also, looking at the code, I see it's always used in value-to-value comparisons - perhaps a method like compareKeys(A, B) would be more appropriate/ clear here?

Reply:
The value returned by Intrinsics.numeric() can be anything numeric. It is just there to make the template code compile. It could be float or double, a constant. Then at code generation time, this intrinsics is simply removed, letting only the actual numerical parameter. E.g `Intrinsics.<KType>numeric(key) > Intrinsics.<KType>numeric(k) ` becomes `key > k` in the generated code.
I should add more javadoc to be clearer.

This intrinsics is used for comparisons but also for other purposes:

Computation - e.g. `int index = (int) (slope * ((double) Intrinsics.<KType>numeric(key) - Intrinsics.<KType>numeric(sKey)) + intercept);` becomes `int index = (int) (slope * ((double) key - sKey) + intercept);`

Method parameter - e.g. `plam.addKey(Intrinsics.<KType>numeric(key), 0, this);` becomes `plam.addKey(key, 0, this);`

I thought that a single Intrinsics.numeric() would be simpler and less invasive than multiple new Intrinsincs. But let me know if we could do differently.

Reply:
Thanks! I wonder if the intrinsic version should cast to an int - perhaps it should use the broadest possible value so that no truncations can occur? I replaced it with a double and all tests seem to pass - made a commit to your branch but feel free to revert if this is incorrect.

Reply:
Thank you Dawid, the javadoc is much clearer.